### PR TITLE
ECM stateless feature optimization

### PR DESCRIPTION
**Pull request by peacewong**
### What is the purpose of the change
ECM stateless feature optimization
### Related issues/PRs

Related issues: #4605,#4606,#4607
Related pr: #4510, #4452

### Brief change log

- Define the core abstraction and interfaces of the EngineConn Factory;
- Define the core abstraction and interfaces of Executor Manager.

### Checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [Linkis mailing list](https://linkis.apache.org/community/how-to-subscribe) first)
- [ ] **If this is a code change**: I have written unit tests to fully verify the new behavior.

### CODE REVIEW

**Comment by CharlieYan24** [comment]
have download code, and finished review, LGTM.

File: linkis-dist/package/sbin/kill-process-by-port.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+if [ "$LINKIS_HOME" = "" ]
+then
+  LINKIS_HOME=$workDir
+fi
+pid=`ps -ef | grep server.port=$port | grep  EngineConnServer | awk '{print $2}'`
+echo "`date '+%Y-%m-%d %H:%M:%S'` Get port $port pid is $pid"
+if [ "$pid" != "" ]
+then
+  sh $LINKIS_HOME/sbin/kill-process-by-pid.sh $pid
+fi
```
Comment (casionone) [inline linkis-dist/package/sbin/kill-process-by-port.sh:28]: it just kill ec service by given port maybe batter   adjust filename 
kill-ec-process-by-port.sh

File: linkis-computation-governance/linkis-manager/linkis-manager-common/src/main/java/org/apache/linkis/manager/common/entity/persistence/ECResourceInfoRecord.java [hunk 2]
```diff
@@ -87,6 +89,14 @@ public String getLabelValue() {
     return labelValue;
   }
 
+  public String getEngineType() {
+    if (StringUtils.isNotBlank(labelValue)) {
+      return labelValue.split(",")[1].split("-")[0];
```
Comment (casionone) [inline linkis-computation-governance/linkis-manager/linkis-manager-common/src/main/java/org/apache/linkis/manager/common/entity/persistence/ECResourceInfoRecord.java:94]: may add  annotation for labelValue eg "hadoop-IDE,spark-2.4.3 "

File: linkis-computation-governance/linkis-engineconn/linkis-engineconn-executor/accessible-executor/src/main/scala/org/apache/linkis/engineconn/callback/service/EngineConnCallback.scala [hunk 3]
```diff
@@ -36,15 +36,15 @@ abstract class AbstractEngineConnStartUpCallback() extends EngineConnCallback wi
     protocol match {
       case protocol: EngineConnStatusCallback =>
         if (protocol.getStatus().equals(NodeStatus.Failed)) {
-          logger.error(s"protocol will send to lm: ${protocol}")
+          logger.error(s"EngineConn Start Failed protocol will send to lm: ${protocol}")
         } else {
```
Comment (casionone) [inline linkis-computation-governance/linkis-engineconn/linkis-engineconn-executor/accessible-executor/src/main/scala/org/apache/linkis/engineconn/callback/service/EngineConnCallback.scala:40]: lm => LM or LinkisManager  
for easy to understand

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/em/DefaultEMEngineService.java [hunk 4]
```diff
@@ -116,11 +114,18 @@ public void stopEngine(EngineNode engineNode, EMNode emNode) {
...
+    ECResourceInfoRecord ecResourceInfo = null;
+    if (StringUtils.isNotBlank(engineNode.getTicketId())) {
+      ecResourceInfo = ecResourceInfoService.getECResourceInfoRecord(engineNode.getTicketId());
+    } else {
+      ecResourceInfo =
+          ecResourceInfoService.getECResourceInfoRecordByInstance(
+              engineNode.getServiceInstance().getInstance());
+    }
+    ecResourceInfoService.getECResourceInfoRecordByInstance(
+        engineNode.getServiceInstance().getInstance());
```
Comment (casionone) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/em/DefaultEMEngineService.java:126]: duplicate code?

File: linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/utils/GovernanceUtils.scala [hunk 5]
```diff
@@ -73,13 +73,41 @@ object GovernanceUtils extends Logging {
...
+  def killProcessByPort(port: String, desc: String, isSudo: Boolean): Unit = {
+    val subProcessKillScriptPath = Configuration.getLinkisHome() + "/sbin/kill-process-by-port.sh"
+    if (
+        StringUtils.isBlank(subProcessKillScriptPath) || !new File(subProcessKillScriptPath)
+          .exists()
+    ) {
+      logger.error(s"Failed to locate kill-script, $subProcessKillScriptPath not exist")
+    } else if (StringUtils.isNotBlank(port)) {
+      val cmd = if (isSudo) {
+        Array("sudo", "sh", subProcessKillScriptPath, String.valueOf(port))
```
Comment (casionone) [inline linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/utils/GovernanceUtils.scala:85]: Input string type about  port   parameter
String.valueOf(port)   can be removed

File: linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/ProcessEngineConnLaunchService.scala [hunk 6]
```diff
@@ -52,110 +48,79 @@ abstract class ProcessEngineConnLaunchService extends AbstractEngineConnLaunchSe
   def setLocalDirsHandleService(localDirsHandleService: LocalDirsHandleService): Unit =
     this.localDirsHandleService = localDirsHandleService
 
-  override def waitEngineConnStart(
+  override def startEngineConnMonitorStart(
       request: EngineConnLaunchRequest,
       conn: EngineConn,
       duration: Long
```
Comment (casionone) [inline linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/ProcessEngineConnLaunchService.scala:54]: ec start timeout parameter  duration  no used? can remove?

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/service/impl/DefaultResourceManager.java [hunk 7]
```diff
@@ -140,20 +140,39 @@ public void register(ServiceInstance serviceInstance, NodeResource resource) {
...
+      Resource usedResource = emResource.getLockedResource().add(emResource.getUsedResource());
+      if (usedResource.moreThan(emResource.getMaxResource())) {
+        logger.warn(
+            "ECM {}  has been registered, but usedResource > MaxResource  need to register resource.",
+            serviceInstance);
+        registerResourceFlag = true;
+      }
+      if (!(resource.getMaxResource().equalsTo(emResource.getMaxResource()))) {
+        logger.warn(
+            "ECM {} has been registered, but Inconsistent newly registered resources  need to register resource.",
```
Comment (casionone) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/service/impl/DefaultResourceManager.java:163]: Inconsistent ->inconsistent

File: linkis-dist/package/sbin/kill-process-by-port.sh (hunk 1 above)
Comment (peacewong) [reply linkis-dist/package/sbin/kill-process-by-port.sh:28]: yes

File: linkis-computation-governance/linkis-manager/linkis-manager-common/src/main/java/org/apache/linkis/manager/common/entity/persistence/ECResourceInfoRecord.java (hunk 2 above)
Comment (peacewong) [reply linkis-computation-governance/linkis-manager/linkis-manager-common/src/main/java/org/apache/linkis/manager/common/entity/persistence/ECResourceInfoRecord.java:94]: ok

File: linkis-computation-governance/linkis-engineconn/linkis-engineconn-executor/accessible-executor/src/main/scala/org/apache/linkis/engineconn/callback/service/EngineConnCallback.scala (hunk 3 above)
Comment (peacewong) [reply linkis-computation-governance/linkis-engineconn/linkis-engineconn-executor/accessible-executor/src/main/scala/org/apache/linkis/engineconn/callback/service/EngineConnCallback.scala:40]: yes

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/em/DefaultEMEngineService.java (hunk 4 above)
Comment (peacewong) [reply linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/em/DefaultEMEngineService.java:126]: Yes need to remove

File: linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/utils/GovernanceUtils.scala (hunk 5 above)
Comment (peacewong) [reply linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/utils/GovernanceUtils.scala:85]: yes

File: linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/ProcessEngineConnLaunchService.scala (hunk 6 above)
Comment (peacewong) [reply linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/ProcessEngineConnLaunchService.scala:54]: Yes

File: linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/java/org/apache/linkis/ecm/server/service/impl/DefaultEngineConnKillService.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+      if (!killEngineConnByPort(processPort, engineStopRequest.getServiceInstance())) {
+        response.setStopStatus(false);
+        response.setMsg(
+            "Kill engine " + engineStopRequest.getServiceInstance().toString() + " failed.");
+      } else {
+        response.setStopStatus(true);
+        response.setMsg(
+            "Kill engine " + engineStopRequest.getServiceInstance().toString() + " succeed.");
+      }
+      response.setStopStatus(false);
```
[the remark this answers is not in the document]
Comment (peacewong) [reply linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/java/org/apache/linkis/ecm/server/service/impl/DefaultEngineConnKillService.java:90]: Need to remove

**Comment by peacewong** [comment]
already updated, ping @casionone

**Review by casionone — APPROVED** [review]

**Comment by casionone** [comment]
LGTM

**Pull request merged**